src/cli/self_update/windows.rs

```diff
@@ -510,83 +504,78 @@ fn _apply_new_path(new_path: Option<Vec<u16>>) -> Result<()> {
-                warn!(
-                    "the registry key HKEY_CURRENT_USER\\Environment\\PATH is not a string. \
-                       Not modifying the PATH variable"
-                );
-                Ok(None)
-            }
+        Ok(val) => Ok(Some(val)),
+        Err(e) if e.code() == HRESULT::from_win32(ERROR_INVALID_DATA) => {
+            warn!(
+                "the registry key HKEY_CURRENT_USER\\Environment\\PATH is not a string. \
+                   Not modifying the PATH variable"
+            );
+            Ok(None)
         }
-        Err(ref e) if e.kind() == io::ErrorKind::NotFound => Ok(Some(Vec::new())),
+        Err(e) if e.code() == HRESULT::from_win32(ERROR_FILE_NOT_FOUND) => Ok(Some(HSTRING::new())),
         Err(e) => Err(e).context(CLIError::WindowsUninstallMadness),
     }
 }
 
 // Returns None if the existing old_path does not need changing, otherwise
 // prepends the path_str to old_path, handling empty old_path appropriately.
-fn _add_to_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
+fn _add_to_path(old_path: HSTRING, path_str: HSTRING) -> Option<HSTRING> {
     if old_path.is_empty() {
         Some(path_str)
     } else if old_path
+        .as_wide()
         .windows(path_str.len())
-        .any(|path| path == path_str)
+        .any(|path| path == path_str.as_wide())
     {
         None
     } else {
-        let mut new_path = path_str;
-        new_path.push(b';' as u16);
-        new_path.extend_from_slice(&old_path);
-        Some(new_path)
+        let mut new_path = path_str.to_os_string();
+        new_path.push(";");
+        new_path.push(old_path.to_os_string());
+        Some(HSTRING::from(new_path))
     }
 }
 
 // Returns None if the existing old_path does not need changing
-fn _remove_from_path(old_path: Vec<u16>, path_str: Vec<u16>) -> Option<Vec<u16>> {
+fn _remove_from_path(old_path: HSTRING, path_str: HSTRING) -> Option<HSTRING> {
     let idx = old_path
+        .as_wide()
         .windows(path_str.len())
-        .position(|path| path == path_str)?;
+        .position(|path| path == path_str.as_wide())?;
     // If there's a trailing semicolon (likely, since we probably added one
     // during install), include that in the substring to remove. We don't search
     // for that to find the string, because if it's the last string in the path,
     // there may not be.
     let mut len = path_str.len();
-    if old_path.get(idx + path_str.len()) == Some(&(b';' as u16)) {
+    if old_path.as_wide().get(idx + path_str.len()) == Some(&(b';' as u16)) {
         len += 1;
     }
 
-    let mut new_path = old_path[..idx].to_owned();
-    new_path.extend_from_slice(&old_path[idx + len..]);
+    let mut new_path = old_path.as_wide()[..idx].to_owned();
+    new_path.extend_from_slice(&old_path.as_wide()[idx + len..]);
     // Don't leave a trailing ; though, we don't want an empty string in the
     // path.
     if new_path.last() == Some(&(b';' as u16)) {
```

Comment:
Although I can convert `HSTRING` to `String` (with some probable losses) to judge which character is the last one in this string, I think adding `ends_with()` method for `HSTRING` seems better.